src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -436,40 +423,103 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+        private bool TryGetRawDataForArray(BoundArrayCreation initializer, out ImmutableArray<byte> data, out int elementCount)
         {
-            data = default;
+            elementCount = GetArrayElementCount(initializer);
 
-            if (initializer == null)
+            if (elementCount < 0)
             {
-                return -1;
+                data = default;
+                return false;
             }
 
-            var initializers = initializer.Initializers;
-            if (initializers.Any(init => init.ConstantValue == null))
+            if (elementCount == 0)
             {
-                return -1;
+                data = ImmutableArray<byte>.Empty;
+                return true;
             }
 
-            var elementCount = initializers.Length;
-            if (elementCount == 0)
+            var arrayType = (ArrayTypeSymbol)initializer.Type;
+            var elementType = arrayType.ElementType.EnumUnderlyingTypeOrSelf();
+
+            if (initializer.InitializerOpt != null)
             {
-                data = ImmutableArray<byte>.Empty;
-                return 0;
+                return TryGetRawDataForArrayInit(initializer.InitializerOpt, elementType, out data);
             }
 
-            var writer = new BlobBuilder(initializers.Length * 4);
+            var elementSize = elementType.SpecialType.SizeInBytes();
+            if (elementSize == 0)
+            {
+                // The type is not a special type and we can not be sure about the element size.
+                data = default;
+                return false;
+            }
+
+            // Since we are dealing with zeroes, we do not care about endianness,
+            // so we create just an array filled with zeroes.
+            var arraySizeInBytes = elementCount * elementSize;
+            data = ImmutableArray.Create(new byte[arraySizeInBytes]);
+            return true;
+
+            static int GetArrayElementCount(BoundArrayCreation initializer)
+            {
+                if (initializer.InitializerOpt != null)
+                {
+                    return initializer.InitializerOpt.Initializers.Length;
+                }
+
+                if (initializer.Bounds == null || initializer.Bounds.Length != 1)
+                {
+                    return -1;
+                }
+
+                var elementCountValue = initializer.Bounds[0].ConstantValue;
+                if (elementCountValue == null)
+                {
+                    return -1;
+                }
+
+                return elementCountValue.Int32Value;
+            }
+        }
+
+        /// <summary>
+        /// Generates a byte blob that matches serialized content of single array initializer.
+        /// </summary>
+        /// <returns>True if the blob is generated successfully, false otherwise</returns>
+        private bool TryGetRawDataForArrayInit(BoundArrayInitialization initializer, TypeSymbol elementType, out ImmutableArray<byte> data)
```

Comment:
Consider either making this a static method or making this a static local function.

Reply:
Fixed.